docs/proposals/provisioning-proposal.md

```diff
@@ -0,0 +1,112 @@
+# Cluster-API Provisioning Mechanism Consolidation Proposal
+
+## Abstract
+
+Cluster-API requires implementers to implement provisioning mechanisms (bootstrap scripts and logic for running those scripts) themselves. However, the mechanism, including bootstrap scripts, is same or very similar for all cloud providers. As this is a time consuming task it could be very useful to provide a reusable provisioning mechanism in the upstream Cluster-API.
+
+## Proposal
+
+The upstream Cluster-API doesn’t have any bootstrap scripts neither any wrapper around kubeadm for handling cluster operations (initialization and join). This means implementers have to manually write bootstrap scripts used by specific Cluster-API provider and manually handle cluster initialization and node joining. For most cloud providers both bootstrap scripts and steps for cluster initialization/joining are the same, except for provider specifics that could be provided via template variables.
+
+The bootstrap scripts usually do the following tasks:
+
+* Install dependencies required by Kubernetes
+    * Includes dependencies such as socat, ebtables, which are same for all providers.
+* Install container runtime
+    * Steps for installing container runtime such as Docker or containerd are same for all providers.
+* Install Kubernetes packages—kubeadm, kubelet, kubectl, kubernetes-cni
+    * The required packages and steps to install them are same for all providers.
+* Configure kubelet
+    * There are several differences based on provider, however differences could be templated.
+    * kubeadm configuration file can handle some configuration tasks for kubelet.
+* Initialize cluster or join node to a cluster
+    * Should be done using kubeadm configuration files, written by implementer or Cluster-API provider user.
+
+Implementing the provisioning scripts in the upstream Cluster-API ensures:
+
+* Cluster-API implementers can easily get started, as they don’t need to implement complex provisioning logic and scripts for many various setups and operating systems
+* The bootstrap scripts are located in one place, which makes it easier to maintain, distribute and test them
+
+### Example use case
+
+Han wants to write a new Cluster-API provider for his employers cloud solution. Han knows the only difference is the cloud providers’ API, hence Han only wants to implement those API calls. Because the Cluster-API provides templates for the actual provisioning, Han can easily import and re-use them and safely assume they will work, as they are well tested.
```

Comment:
@roberthbailey: How will we test that a change upstream to these scripts will work on all providers next time they vendor the latest upstream repo? And I'm not sure Han can assume they will "just work" because with required templating Han will need to pass values that are unique to the cloud environment.

@alvaroaleman: So IMHO what we can test on the cluster-api main repo is that these scripts work when `--cloud-provider` and `--cloud-config` is unset, e.G. by leveraging KIND and then just creating a node inside a docker container via DIND and checking if it will come up and join the cluster. Aside from those two settings, I think the node config is identical on all cloud providers or am I missing something?